libtransmission/peer-mgr.c

```diff
@@ -2800,8 +2803,31 @@ struct tr_peer_stat* tr_peerMgrPeerStats(tr_torrent const* tor, int* setmeCount)
         stat->progress = peer->progress;
         stat->isUTP = tr_peerMsgsIsUtpConnection(msgs);
         stat->isEncrypted = tr_peerMsgsIsEncrypted(msgs);
-        stat->rateToPeer_KBps = toSpeedKBps(tr_peerGetPieceSpeed_Bps(peer, now_msec, TR_CLIENT_TO_PEER));
-        stat->rateToClient_KBps = toSpeedKBps(tr_peerGetPieceSpeed_Bps(peer, now_msec, TR_PEER_TO_CLIENT));
+        // The rates jump around a lot, so smooth them out. We are considering this a data collection issue rather than
+        // presentation issue, and so are smoothing here rather than within the respective UIs.
+        double newRateToPeer_KBps = toSpeedKBps(tr_peerGetPieceSpeed_Bps(peer, now_msec, TR_CLIENT_TO_PEER));
+        if (newRateToPeer_KBps > stat->rateToPeer_KBps)
+        {
+            stat->rateToPeer_KBps = stat->rateToPeer_KBps + ((newRateToPeer_KBps - stat->rateToPeer_KBps) / SMOOTHING_RATE_UP);
+        }
+        else
+        {
+            stat->rateToPeer_KBps = stat->rateToPeer_KBps +
+                ((newRateToPeer_KBps - stat->rateToPeer_KBps) / SMOOTHING_RATE_DOWN);
+        }
+
+        double newRateToClient_KBps = toSpeedKBps(tr_peerGetPieceSpeed_Bps(peer, now_msec, TR_PEER_TO_CLIENT));
+        if (newRateToClient_KBps > stat->rateToClient_KBps)
+        {
+            stat->rateToClient_KBps = stat->rateToClient_KBps +
+                ((newRateToClient_KBps - stat->rateToClient_KBps) / SMOOTHING_RATE_UP);
+        }
+        else
+        {
+            stat->rateToClient_KBps = stat->rateToClient_KBps +
+                ((newRateToClient_KBps - stat->rateToClient_KBps) / SMOOTHING_RATE_DOWN);
+        }
```

Comment:
- This will be a little strange when it shows that we're still downloading a torrent even after the torrent is complete.

- We probably shouldn't smooth when the previous value is zero.

- There is some minor code duplication here. The repetition of the same similar variable names makes this code hard to scan. Maybe extract it out into a helper function / lambda?